### ISSUE TITLE
There is no possiblity of upgrading TB using internal updater?

### ISSUE THREAD

**Issue by TNTBOMBOM** [flathub/org.torproject.torbrowser-launcher#5]
How can TB upgrade itself while it has no ability to do that from within the browser? (even notification of new version its not shown, i need to do that all manually)

**Comment by AsciiWolf** [comment]
This is a [known issue](https://github.com/torproject/torbrowser-launcher/issues/721) that is being worked on and will hopefully be fixed soon. However, your question is valid - it will sadly require existing users to wipe the app data and reinstall Tor Browser. :-/

Anyway, please take a look at this ticket: https://gitlab.torproject.org/tpo/applications/tor-browser/-/issues/42293

**Comment by mooreye** [comment]
> This is a [known issue](https://github.com/torproject/torbrowser-launcher/issues/721) that is being worked on and will hopefully be fixed soon. However, your question is valid - it will sadly require existing users to wipe the app data and reinstall Tor Browser. :-/
> 
> Anyway, please take a look at this ticket: https://gitlab.torproject.org/tpo/applications/tor-browser/-/issues/42293

I need some clarification.

The linked issue is closed. I see that an env var `TORBROWSER_LAUNCHER` was introduced but it doesn't fix anything (it is reserved for future use, if you will). The codebase of TB Launcher was moved from GitHub to TB Gitlab.

* is this the official Flatpak version of TB that was supposed to be made?
* what is being done to fix the non-updating issue?
* will there be yet another Flatpak version of TB that does not include a downloader and instead will contain TB's latest version as part of an actual tagged Flathub release, therefore removing the need for fixing the in-browser updater altogether?
* for now, is it better to just use the regular Tor Browser downloaded directly from TB Project's website?

**Comment by AsciiWolf** [comment]
It should be fixed when Tor Browser 13.0.10 gets released. This will hopefully happen soon (see [this ticket](https://gitlab.torproject.org/tpo/applications/tor-browser-build/-/issues/41061)). Unfortunately, current torbrowser-launcher Flatpak users will need to remove the existing data directory (can be done via GUI when uninstalling torbrowser-launcher via GNOME Software / KDE Discover) to trigger the latest Tor Browser archive download.

**Comment by pm4rcin** [comment]
@AsciiWolf there's already Tor Browser 13.0.11. Is something blocking updating or you simply had no time/forgot?

**Comment by mooreye** [comment]
it works fine but redownload of the TB Bundle is needed.

**Comment by AsciiWolf** [comment]
Yep, it works, but sadly, there is no other solution available for existing users than deleting the Flatpak data to trigger the TBB re-download.

**Comment by AsciiWolf** [comment]
Yes, that is a possible workaround.

**Comment by Francewhoa** [comment]
If needed, both me and the Ubertus.Org team would be happy to contribute testing and documentation for upgrading TB using internal updater. 

Any volunteer to implement this new feature?

**Comment by AsciiWolf** [comment]
There is a working internal updater, problem is that it was added in TBB 13.0.10. I am thinking about adding some mechanism to detect TBB version directly by torbrowser-launcher and updating it if it is too old, however my free time is sadly very limited. :-/ The current transition process for current users sadly is to manually wipe the Tor Browser Launcher Flatpak data to trigger a clean, latest TBB package download. Everything should work fine from that point.

**Comment by Francewhoa** [comment]
About the temporary workaround to either deleting the Flatpak data to trigger the Tor Browser Bundle (TBB) re-download or using Tor Browser Launcher (TBL) to reinstall TB. This works great to install or reinstall a fresh new TBB. But from the end-user's point of view, this workaround is not an update. This is just a plain delete TBB and its configuration, then reinstall a new TBB. Update and reinstall are not to be confused.

I mean, for most end-users, the word "update" means that after the TBB update their TBB configuration remains. But, as you know, presently, as of April 12th, 2024, TBL is not yet able to both update TBB and keep the user's TBB configuration. I understand that adding an automated update feature to TBL is in development.

About this Flathub page at https://flathub.org/apps/org.torproject.torbrowser-launcher The challenge is that it claims that Tor Browser Launcher (TBL) version 0.3.7 is able to automatically update Tor Browser Bundle (TBB). 

To resolve this challenge above, I suggest this new ticket https://github.com/flathub/org.torproject.torbrowser-launcher/issues/19

**Comment by Francewhoa** [comment]
Thanks @AsciiWolf :) For both your reply and your contribution to Tor Browser Launcher (TBL) on Flathub

-----------

> The current transition process for current users sadly is to manually wipe the Tor Browser Launcher Flatpak data to trigger a clean, latest TBB package download. Everything should work fine from that point.

This fails with the present TBL version 0.3.7. In other words, it does not work fine from that point.

Steps to reproduce:
1. Install TBL 0.3.7 from https://flathub.org/apps/org.torproject.torbrowser-launcher
1. Using TBL 0.3.7,

**Comment by Francewhoa** [comment]
Thanks @AsciiWolf :) For both your reply and your contribution to Tor Browser Launcher (TBL) on Flathub

-----------

> The current transition process for current users sadly is to manually wipe the Tor Browser Launcher Flatpak data to trigger a clean, latest TBB package download. Everything should work fine from that point.

This is to confirm this always fails with the present TBL version 0.3.7. In other words, it does not work fine from that point.

**Steps to reproduce:**
1. Install TBL 0.3.7 from https://flathub.org/apps/org.torproject.torbrowser-launcher
1. Using TBL 0.3.7, install the latest Tor Browser Bundle (TBB). TBL is able to successfully do this :)
1. Using the TBB, change its configuration. For example, but not limited to, add a new bookmark.
1. Using TBL 0.3.7, try to update Tor Browser Bundle (TBB) while keeping its configuration, including, but not limited to keeping the new bookmark you just created above. This update always fails. This is the challenge. By "fails" we mean TBL does update TBB, but does not keep its configuration. Instead, TBL deleted TBB configuration. The needed end results it that TBL update TBB while keeping its present configuration.

For most end-user,  the word "update" means that after the TBB update, their TBB configuration remains. But, per the steps to reproduce above, TBL is not yet able to both update TBB and keep the user's TBB configuration. What TBL presently really does is fully delete both the TBB and delete the user's TBB configuration, then reinstall a fresh TBB.

What most users need is this:

1. First, use TBL to install TBB. Then, configuring their TBB. Some user spending hours installing and configuring modules/extensions, theme, toolbars, etc. TBL is presently successful at this.
1. Then, automatically be notified when a new TBB update is available. Then somehow be able to update it. Either via TBL or TBB. TBL is not presently able to do this.
1. Next, after the above update, their configuration remains. TBL is not presently able to do this.

Any volunteer to add this new feature so that TBL is able to both update TBB and keep its configuration?

If needed, both me and the Ubertus.Org team would be happy to contribute testing and documentation for upgrading TBB using internal updater, while keeping its configuration

**Comment by boklm** [comment]
> > The current transition process for current users sadly is to manually wipe the Tor Browser Launcher Flatpak data to trigger a clean, latest TBB package download. Everything should work fine from that point.
> 
> This is to confirm this always fails with the present TBL version 0.3.7. In other words, it does not work fine from that point.

Which version of Tor Browser do you have installed?

There was a bug that disabled the updater when running
torbrowser-launcher flatpak, but this was fixed with Tor Browser 13.0.10
and torbrowser-launcher 0.3.7.

If you have Tor Browser 13.0.10 or newer installed, the updater is
expected to work. If it still doesn't work then this is a different bug.

**Comment by Francewhoa** [comment]
Thanks for your reply @boklm :) My replies are below. Including two questions for you down below.

------------------

> Which version of Tor Browser do you have installed?

Tor Browser (TB) 13.0.9 (based on Mozilla Firefox 115.7.0esr). Which was installed with Tor Browser Launcher (TBL) 0.3.7. This screenshot shows that within TB, the TB updater somehow failed. It is not displayed at all. So I'm assuming it did not check for update.

![TBB---13 0 9---Failed_update](https://github.com/flathub/org.torproject.torbrowser-launcher/assets/119942/fa9a6840-0c20-49e7-8b4b-bfd53023a3a2)

------------------

Just now, we tried using Tor Browser Launcher (TBL) 0.3.7 to reinstall TB 13.0.14 (Based on Mozilla Firefox 115.10.0esr). This screenshot shows that the updater successfully checked for update. Yayaya :) This screenshot is in French. As you know, the green checkmark means the update check was successful.

![TBB---13 0 14---Success_check_for_update](https://github.com/flathub/org.torproject.torbrowser-launcher/assets/119942/ed77d725-9785-4c85-9a81-9caf884a9706)

After using Tor Browser Launcher (TBL) 0.3.7 reinstalled TB 13.0.14 over an already installed and configured TB 13.0.9, all the user's TB configurations were lost :(

As for the TB update using TB, in the future, after the next TB version is released, we will know if this TB updater work

---------------

Question 1.  In the future, if the user needs to keep its already installed TB configuration, we need to use TB to update TB. Correct? Not use TBL to update TB.

Question 2. Assuming the TB was installed with TBL. TBL itself, alone, and independently, is not able to update an already installed TB, and keep the user's TB configuration. Because TB updates are done within TB itself. In other words, if TB was installed with TBL, after that, TB is able to both update itself and keep the user's configuration. Correct?

**Comment by boklm** [comment]
On Wed, 17 Apr 2024, Francois Carpentier wrote:

> Thanks for your reply @boklm :) My replies are below. Including two questions for you down below.
> 
> ------------------
> 
> > Which version of Tor Browser do you have installed?
> 
> Tor Browser (TB) 13.0.9 (based on Mozilla Firefox 115.7.0esr). Which was installed with Tor Browser Launcher (TBL) 0.3.7. This screenshot shows that within TB, the TB updater somehow failed. It is not displayed at all. So I'm assuming it did not check for update.
> 
> ![TBB---13 0 9---Failed_update](https://github.com/flathub/org.torproject.torbrowser-launcher/assets/119942/fa9a6840-0c20-49e7-8b4b-bfd53023a3a2)

Yes, this is expected with Tor Browser 13.0.9.

However the issue only happens when running torbrowser-launcher with the
flatpak. If I remember correctly a possible workaround is to run the
browser without using flatkpak (just one time), which should allow the
browser to update itself.

> ---------------
> 
> Question 1.  In the future, if the user needs to keep its already installed TB configuration, we need to use TB to update TB. Correct? Not use TBL to update TB.

Yes.

> Question 2. Assuming the TB was installed with TBL. TBL itself, alone, and independently, is not able to update an already installed TB, and keep the user's TB configuration. Because TB updates are done within TB itself. In other words, if TB was installed with TBL, after that, TB is able to both update itself and keep the user's configuration. Correct? 

Yes.

**Comment by Francewhoa** [comment]
> If I remember correctly a possible workaround is to run the
> browser without using flatkpak (just one time), which should allow the
> browser to update itself.

@boklm, this is to confirm this worked. Thanks :) We restored the user's TB configuration from a backup. Then, after the workaround, the configuration remains.

------------------

### For those not familiar with this workaround, the steps are:

1. Close Tor Browser (TB)

1. Using Terminal/Console, execute this command as user to change to the appropriate directory. Where the TB was installed to by Tor Browser Launcher (TBL) for example, for Debian, the command is:

    `cd /home/<DEBIAN USER NAME>/.var/app/org.torproject.torbrowser-launcher/data/torbrowser/tbb/x86_64/tor-browser/Browser/`

     * Note:
     
         * In this command above, you need to replace `<DEBIAN USER NAME>` for your Linux user name
         * This command need to be executed as user. So without sudo and without root.
         
1. Using Terminal, execute this command to launch TB outside Flatpak:

    `./start-tor-browser`
  
     * Note:
     
         * The prefix `./` is required
         * This command need to be executed as user. So without sudo and without root.
         
1. TB will open. Using TB, navigate to  [Options] » button ---> « Help » option --->  « About Tor Browser » 

1. TB will automatically check for update and down it if any. This screenshot shows this.
  
   ![TBB---13 0 14---One_time_outside_flatpak---Success_updated---1](https://github.com/flathub/org.torproject.torbrowser-launcher/assets/119942/0f730987-3387-4c7b-947d-4ddee250ad81)

1. After the update is fully downloaded, close this « About Tor Browser ». Do not click on "Restart" button.
   ![TBB---13 0 14---One_time_outside_flatpak---Success_updated---2](https://github.com/flathub/org.torproject.torbrowser-launcher/assets/119942/34fdf6c7-4d7b-4ab5-95d4-c9166a574bc1)
   
1. Using GNOME, launch TB. It will automatically start with Flatpak. All TB previous configuration successfully remains.

1. In the future, to update TB, using TB, navigate to  [Options] » button ---> « Help » option --->  « About Tor Browser ». After the update is fully downloaded, click on "Restart" button.

**Comment by AsciiWolf** [comment]
Yep, this is also a possible workaround, in fact it is the one I used on my systems. Thanks for documenting these steps!
